pkg/rules/service.go

```diff
@@ -0,0 +1,347 @@
+	K8s       coreV1.CoreV1Interface
+}
+
+type Service struct {
+	oathkeeper oathkeeper.ApiApi
+
+	cmName      string
+	cmNamespace string
+	ruleFiles   []string
+
+	k8s coreV1.CoreV1Interface
+
+	tracer  trace.Tracer
+	monitor monitoring.MonitorInterface
+	logger  logging.LoggerInterface
+}
+
+func (s *Service) ListRules(ctx context.Context, limit int64, offset int64) ([]*RuleData, error) {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.ListRules")
+	defer span.End()
+
+	rules, _, err := s.oathkeeper.ListRules(ctx).Limit(limit).Offset(offset).Execute()
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	ruleDataList := make([]*RuleData, 0)
+
+	rawRules, err := json.Marshal(rules)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal([]byte(rawRules), &ruleDataList)
+
+	if err != nil {
+		s.logger.Errorf("failed converting oathkeeper Rule list to RuleData list %s - %v", rawRules, err)
+		return nil, err
+	}
+
+	return ruleDataList, nil
+}
+
+func (s *Service) GetRule(ctx context.Context, id string) (*RuleData, error) {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.GetRule")
+	defer span.End()
+
+	rule, _, err := s.oathkeeper.GetRule(ctx, id).Execute()
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return nil, err
+	}
+
+	ruleDataConversion := RuleData{}
+
+	rawRule, err := json.Marshal(rule)
+
+	if err != nil {
+		return nil, err
+	}
+
+	err = json.Unmarshal([]byte(rawRule), &ruleDataConversion)
+
+	if err != nil {
+		s.logger.Errorf("failed converting oathkeeper Rule to RuleData object %s - %v", rawRule, err)
+		return nil, err
+	}
+
+	return &ruleDataConversion, nil
+}
+
+func (s *Service) UpdateRule(ctx context.Context, id string, newRule RuleData) error {
+	ctx, span := s.tracer.Start(ctx, "rules.Service.UpdateRule")
+	defer span.End()
+
+	cm, err := s.k8s.ConfigMaps(s.cmNamespace).Get(ctx, s.cmName, metaV1.GetOptions{})
+
+	if err != nil {
+		s.logger.Error(err.Error())
+		return err
+	}
+
+	if s.ruleFiles == nil {
+		keys := make([]string, 0)
+		for k := range cm.Data {
+			keys = append(keys, k)
+		}
+		if len(keys) > 0 {
+			s.ruleFiles = keys
+		} else {
+			return fmt.Errorf("rule with ID %s not found", id)
+		}
+	}
+
+	for _, fileName := range s.ruleFiles {
+		rules := s.extractRules(cm.Data, fileName)
+
+		if rules == nil {
+			continue
+		}
+
+		var checkRuleIncluded *RuleData
+
+		for i, r := range rules {
+			if r.Id == id {
+				rules[i] = &newRule
+				checkRuleIncluded = r
+			}
+		}
+
+		if checkRuleIncluded == nil {
+			continue
+		}
+
+		rawRules, err := json.Marshal(rules)
+
+		if err != nil {
+			return err
+		}
+
+		cm.Data[fileName] = string(rawRules)
+
+		if _, err = s.k8s.ConfigMaps(s.cmNamespace).Update(ctx, cm, metaV1.UpdateOptions{}); err != nil {
+
+			return err
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("rule with ID %s not found", id)
+}
+
+func (s *Service) CreateRule(ctx context.Context, id string, newRule RuleData, userFile string) error {
```

Comment:
I think that it would make our life easier if we created a different file for each user created rule. That way we wouldn't have to iterate over all the files every time. 

We can discuss it after standup

Reply:
I agree. Done.